File: app/src/main/java/ua/hospes/nfs/marathon/data/sessions/SessionsRepositoryImpl.java
package ua.hospes.nfs.marathon.data.sessions;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Observable;
import rx.Subscriber;
import ua.hospes.nfs.marathon.data.sessions.mapper.SessionsMapper;
import ua.hospes.nfs.marathon.data.sessions.models.SessionDb;
import ua.hospes.nfs.marathon.data.sessions.operations.CloseSessionOperation;
import ua.hospes.nfs.marathon.data.sessions.operations.OpenSessionOperation;
import ua.hospes.nfs.marathon.data.sessions.operations.SetCarOperation;
import ua.hospes.nfs.marathon.data.sessions.operations.SetDriverOperation;
import ua.hospes.nfs.marathon.data.sessions.storage.SessionsDbStorage;
import ua.hospes.nfs.marathon.domain.cars.CarsRepository;
import ua.hospes.nfs.marathon.domain.cars.models.Car;
import ua.hospes.nfs.marathon.domain.drivers.DriversRepository;
import ua.hospes.nfs.marathon.domain.drivers.models.Driver;
import ua.hospes.nfs.marathon.domain.sessions.SessionsRepository;
import ua.hospes.nfs.marathon.domain.sessions.models.Session;

/**
 * @author Andrew Khloponin
 */
@Singleton
public class SessionsRepositoryImpl implements SessionsRepository {
    private final SessionsDbStorage dbStorage;
    private final DriversRepository driversRepository;
    private final CarsRepository carsRepository;


    @Inject
    public SessionsRepositoryImpl(SessionsDbStorage dbStorage, DriversRepository driversRepository, CarsRepository carsRepository) {
        this.dbStorage = dbStorage;
        this.driversRepository = driversRepository;
        this.carsRepository = carsRepository;
    }


    @Override
    public Observable<Session> get() {
        return dbStorage.get()
                .flatMap(sessionItemDb ->
                        Observable.zip(
                                Observable.just(sessionItemDb),
                                getDriverById(sessionItemDb.getDriverId()),
                                getCarById(sessionItemDb.getCarId()),
                                SessionsMapper::map));
    }

    @Override
    public Observable<Session> get(int... id) {
        return dbStorage.get(id)
                .flatMap(sessionItemDb ->
                        Observable.zip(
                                Observable.just(sessionItemDb),
                                getDriverById(sessionItemDb.getDriverId()),
                                getCarById(sessionItemDb.getCarId()),
                                SessionsMapper::map));
    }

    @Override
    public Observable<Session> getByTeamId(int teamId) {
        return dbStorage.getByTeamId(teamId)
                .flatMap(sessionItemDb ->
                        Observable.zip(
                                Observable.just(sessionItemDb),
                                getDriverById(sessionItemDb.getDriverId()),
                                getCarById(sessionItemDb.getCarId()),
                                SessionsMapper::map));
    }

    @Override
    public Observable<Session> getByTeamIdAndDriverId(int teamId, int driverId) {
        return dbStorage.getByTeamIdAndDriverId(teamId, driverId)
                .flatMap(sessionItemDb ->
                        Observable.zip(
                                Observable.just(sessionItemDb),
                                getDriverById(sessionItemDb.getDriverId()),
                                getCarById(sessionItemDb.getCarId()),
                                SessionsMapper::map));
    }

    @Override
    public Observable<List<Session>> listen() {
        return dbStorage.listen()
                .flatMap(sessionItemDbs -> Observable.from(sessionItemDbs)
                        .flatMap(sessionItemDb ->
                                Observable.zip(
                                        Observable.just(sessionItemDb),
                                        getDriverById(sessionItemDb.getDriverId()),
                                        getCarById(sessionItemDb.getCarId()),
                                        SessionsMapper::map))
                        .toList());
    }

    @Override
    public Observable<List<Session>> listenByTeamId(int teamId) {
        return dbStorage.listenByTeamId(teamId)
                .flatMap(sessionItemDbs -> Observable.from(sessionItemDbs)
                        .flatMap(sessionItemDb ->
                                Observable.zip(
                                        Observable.just(sessionItemDb),
                                        getDriverById(sessionItemDb.getDriverId()),
                                        getCarById(sessionItemDb.getCarId()),
                                        SessionsMapper::map))
                        .toList());
    }

    @Override
    public Observable<Session> newSessions(Session.Type type, int... teamIds) {
        return Observable.create(new Observable.OnSubscribe<SessionDb>() {
            @Override
            public void call(Subscriber<? super SessionDb> subscriber) {
                for (int teamId : teamIds) {
                    SessionDb sessionDb = new SessionDb(teamId);
                    sessionDb.setType(type.name());
                    subscriber.onNext(sessionDb);
                }
                subscriber.onCompleted();
            }
        })
                .toList()
                .flatMap(dbStorage::add)
                .flatMap(result ->
                        Observable.zip(
                                Observable.just(result.getData()),
                                getDriverById(result.getData().getDriverId()),
                                getCarById(result.getData().getCarId()),
                                SessionsMapper::map)
                );
    }

    @Override
    public Observable<Session> setSessionDriver(int sessionId, int driverId) {
        return Observable.just(new SetDriverOperation(sessionId, driverId))
                .toList()
                .flatMap(dbStorage::applySetDriverOperations)
                .flatMap(this::get);
    }

    @Override
    public Observable<Session> setSessionCar(int sessionId, int carId) {
        return Observable.just(new SetCarOperation(sessionId, carId))
                .toList()
                .flatMap(dbStorage::applySetCarOperations)
                .flatMap(this::get);
    }

    @Override
    public Observable<Session> startSessions(long startTime, int... sessionIds) {
        return Observable.create(new Observable.OnSubscribe<OpenSessionOperation>() {
            @Override
            public void call(Subscriber<? super OpenSessionOperation> subscriber) {
                for (int sessionId : sessionIds) {
                    subscriber.onNext(new OpenSessionOperation(sessionId, startTime));
                }
                subscriber.onCompleted();
            }
        })
                .toList()
                .flatMap(dbStorage::applyOpenOperations)
                .flatMap(this::get);
    }

    @Override
    public Observable<Session> startNewSessions(long startTime, Session.Type type, int... teamIds) {
        return Observable.create(new Observable.OnSubscribe<SessionDb>() {
            @Override
            public void call(Subscriber<? super SessionDb> subscriber) {
                for (int teamId : teamIds) {
                    SessionDb sessionDb = new SessionDb(teamId);
                    sessionDb.setStartDurationTime(startTime);
                    sessionDb.setType(type.name());
                    subscriber.onNext(sessionDb);
                }
                subscriber.onCompleted();
            }
        })
                .toList()
                .flatMap(dbStorage::add)
                .flatMap(result ->
                        Observable.zip(
                                Observable.just(result.getData()),
                                getDriverById(result.getData().getDriverId()),
                                getCarById(result.getData().getCarId()),
                                SessionsMapper::map)
                );
    }

    @Override
    public Observable<Session> startNewSession(long startTime, Session.Type type, int driverId, int teamId) {
        return Observable.create(new Observable.OnSubscribe<SessionDb>() {
            @Override
            public void call(Subscriber<? super SessionDb> subscriber) {
                SessionDb sessionDb = new SessionDb(teamId);
                sessionDb.setStartDurationTime(startTime);
                sessionDb.setDriverId(driverId);
                sessionDb.setType(type.name());
                subscriber.onNext(sessionDb);
                subscriber.onCompleted();
            }
        })
                .toList()
                .flatMap(dbStorage::add)
                .flatMap(result ->
                        Observable.zip(
                                Observable.just(result.getData()),
                                getDriverById(result.getData().getDriverId()),
                                getCarById(result.getData().getCarId()),
                                SessionsMapper::map)
                );
    }

    @Override
    public Observable<Session> closeSessions(long stopTime, int... sessionIds) {
        return Observable.create(new Observable.OnSubscribe<CloseSessionOperation>() {
            @Override
            public void call(Subscriber<? super CloseSessionOperation> subscriber) {
                for (int sessionId : sessionIds) {
                    subscriber.onNext(new CloseSessionOperation(sessionId, stopTime));
                }
                subscriber.onCompleted();
            }
        })
                .toList()
                .flatMap(dbStorage::applyCloseOperations)
                .flatMap(this::get);
    }


    @Override
    public Observable<Void> clean() {
        return dbStorage.clean();
    }


    private Observable<Driver> getDriverById(int id) {
        return driversRepository.get(id).singleOrDefault(null, driver -> id == driver.getId());
    }

    private Observable<Car> getCarById(int id) {
        return carsRepository.getByIds(id).singleOrDefault(null, car -> id == car.getId());
    }
}

File: app/src/main/java/ua/hospes/rtm/data/SessionsRepositoryImpl.kt
package ua.hospes.rtm.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import ua.hospes.rtm.db.AppDatabase
import ua.hospes.rtm.db.cars.CarDAO
import ua.hospes.rtm.db.drivers.DriverDAO
import ua.hospes.rtm.db.sessions.SessionDAO
import ua.hospes.rtm.db.sessions.SessionEntity
import ua.hospes.rtm.db.sessions.toDomain
import ua.hospes.rtm.db.team.TeamDAO
import ua.hospes.rtm.domain.sessions.Session
import ua.hospes.rtm.domain.sessions.SessionsRepository

internal class SessionsRepositoryImpl(db: AppDatabase) : SessionsRepository {
    private val dao: SessionDAO = db.sessionDao()
    private val teamDAO: TeamDAO = db.teamDao()
    private val driverDAO: DriverDAO = db.driverDao()
    private val carDAO: CarDAO = db.carDao()
    private val raceDAO = db.raceDao()

    override suspend fun get(): List<Session> = withContext(Dispatchers.IO) {
        dao.get().map { it.toDomain(teamDAO, driverDAO, carDAO) }
    }

    override suspend fun get(vararg ids: Long): List<Session> = withContext(Dispatchers.IO) {
        dao.get(*ids).map { it.toDomain(teamDAO, driverDAO, carDAO) }
    }

    override suspend fun getByTeam(teamId: Long): List<Session> = withContext(Dispatchers.IO) {
        dao.getByTeam(teamId).map { it.toDomain(teamDAO, driverDAO, carDAO) }
    }

    override suspend fun getByTeamAndDriver(teamId: Long, driverId: Long): List<Session> = withContext(Dispatchers.IO) {
        dao.getByTeamAndDriver(teamId, driverId).map { it.toDomain(teamDAO, driverDAO, carDAO) }
    }


    override fun listen(): Flow<List<Session>> =
            dao.observe().map { list -> list.map { it.toDomain(teamDAO, driverDAO, carDAO) } }

    override fun listenByTeamId(teamId: Long): Flow<List<Session>> =
            dao.observeByTeam(teamId).map { list -> list.map { it.toDomain(teamDAO, driverDAO, carDAO) } }

    override fun listenByRaceId(raceId: Long): Flow<List<Session>> =
            dao.observeByRace(raceId).map { list -> list.map { it.toDomain(teamDAO, driverDAO, carDAO) } }


    override suspend fun setSessionDriver(sessionId: Long, driverId: Long) =
            withContext(Dispatchers.IO) { dao.setDriver(sessionId, driverId) }

    override suspend fun setSessionCar(sessionId: Long, carId: Long) =
            withContext(Dispatchers.IO) { dao.setCar(sessionId, carId) }

    override suspend fun startRace(time: Long) = withContext(Dispatchers.IO) { dao.startRace(time) }
    override suspend fun stopRace(time: Long) = withContext(Dispatchers.IO) { dao.stopRace(time) }
    override suspend fun resetRace() = withContext(Dispatchers.IO) { dao.resetRace() }

    override suspend fun newSession(type: Session.Type, teamId: Long): Session = withContext(Dispatchers.IO) {
        val id = dao.save(SessionEntity(teamId = teamId, type = type.name))
        Session(id = id, teamId = teamId, type = type)
    }

    override suspend fun newSessions(type: Session.Type, vararg teamIds: Long) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override suspend fun startSessions(raceStartTime: Long, startTime: Long, vararg sessionIds: Long) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override suspend fun startNewSessions(raceStartTime: Long, startTime: Long, type: Session.Type, vararg teamIds: Long) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override suspend fun startNewSession(teamId: Long, raceStartTime: Long, startTime: Long, type: Session.Type, driverId: Long?): Session {
        val entity = SessionEntity(teamId = teamId,
                raceStartTime = raceStartTime,
                startDurationTime = startTime,
                type = type.name)
        val id = dao.save(entity)
        return entity.copy(id = id).toDomain(teamDAO, driverDAO, carDAO)
    }

    override suspend fun closeSessions(stopTime: Long, vararg sessionIds: Long) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override suspend fun removeLastSession(teamId: Long) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override suspend fun clear() = withContext(Dispatchers.IO) { dao.clear() }


    //    override fun newSessions(type: Session.Type, vararg teamIds: Int): Observable<Session> =
    //            Observable.create { subscriber: ObservableEmitter<SessionEntity> ->
    //                for (teamId in teamIds) {
    //                    val sessionDb = SessionEntity(teamId)
    //                    sessionDb.type = type.name
    //                    subscriber.onNext(sessionDb)
    //                }
    //                subscriber.onComplete()
    //            }
    //                    .toList()
    //                    .flatMapObservable { dbStorage.add(it) }.map { it.data }
    //                    .flatMapSingle { transform(it) }
    //
    //    override fun startSessions(raceStartTime: Long, startTime: Long, vararg sessionIds: Int): Observable<Session> =
    //            OpenSessionOperation.from(raceStartTime, startTime, *sessionIds)
    //                    .toList()
    //                    .flatMapObservable { dbStorage.applyOpenOperations(it) }
    //                    .flatMap { get(it) }
    //
    //    override fun startNewSessions(raceStartTime: Long, startTime: Long, type: Session.Type, vararg teamIds: Int): Observable<Session> =
    //            Observable.create { subscriber: ObservableEmitter<SessionEntity> ->
    //                for (teamId in teamIds) {
    //                    val sessionDb = SessionEntity(teamId)
    //                    sessionDb.raceStartTime = raceStartTime
    //                    sessionDb.startDurationTime = startTime
    //                    sessionDb.type = type.name
    //                    subscriber.onNext(sessionDb)
    //                }
    //                subscriber.onComplete()
    //            }
    //                    .toList()
    //                    .flatMapObservable { dbStorage.add(it) }.map { it.data }
    //                    .flatMapSingle { transform(it) }
    //
    //    override fun startNewSession(raceStartTime: Long, startTime: Long, type: Session.Type, driverId: Int, teamId: Int): Observable<Session> =
    //            Observable.create { subscriber: ObservableEmitter<SessionEntity> ->
    //                val sessionDb = SessionEntity(teamId)
    //                sessionDb.raceStartTime = raceStartTime
    //                sessionDb.startDurationTime = startTime
    //                sessionDb.driverId = driverId
    //                sessionDb.type = type.name
    //                subscriber.onNext(sessionDb)
    //                subscriber.onComplete()
    //            }
    //                    .toList()
    //                    .flatMapObservable { dbStorage.add(it) }.map { it.data }
    //                    .flatMapSingle { transform(it) }
    //
    //    override fun closeSessions(stopTime: Long, vararg sessionIds: Int): Observable<Session> =
    //            CloseSessionOperation.from(stopTime, *sessionIds)
    //                    .toList()
    //                    .flatMapObservable { dbStorage.applyCloseOperations(it) }
    //                    .flatMap { get(it) }
    //
    //    override fun removeLastSession(teamId: Int): Observable<Session> =
    //            dbStorage.getByTeamId(teamId)
    //                    .toSortedList { o1, o2 -> o1.startDurationTime.compareTo(o2.startDurationTime) }
    //                    .flatMapObservable { list ->
    //                        val count = list.size
    //                        when {
    //                            count >= 2 -> dbStorage.getByTeamId(teamId)
    //                                    .toSortedList { o1, o2 -> o1.startDurationTime.compareTo(o2.startDurationTime) }
    //                                    .flatMapObservable {
    //                                        Single.zip(
    //                                                zipOpenSession(it[count - 2]),
    //                                                removeSession(it[count - 1]),
    //                                                BiFunction { t1: SessionEntity, _: Int -> t1 }
    //                                        ).toObservable()
    //                                    }
    //
    //                            count == 1 ->
    //                                dbStorage.getByTeamId(teamId)
    //                                        .toSortedList { o1, o2 -> o1.startDurationTime.compareTo(o2.startDurationTime) }
    //                                        .flatMapObservable { Observable.just(it.first()) }
    //                            else -> throw RuntimeException("No sessions in the database")
    //                        }
    //                    }
    //                    .flatMapSingle { transform(it) }
    //
    //    private fun zipOpenSession(sessionDb: SessionEntity): Single<SessionEntity> = Single.zip(
    //            Single.just(sessionDb),
    //            openSession(sessionDb.id, sessionDb.raceStartTime, sessionDb.startDurationTime),
    //            BiFunction { sessionDb1, _ -> sessionDb1 }
    //    )
    //
    //    private fun openSession(id: Int, raceStartTime: Long, startTime: Long): Single<Int> =
    //            Observable.just(OpenSessionOperation(id, raceStartTime, startTime))
    //                    .toList()
    //                    .flatMapObservable { dbStorage.applyOpenOperations(it) }
    //                    .single(0)
    //
    //    private fun removeSession(sessionDb: SessionEntity): Single<Int> = dbStorage.remove(sessionDb)
}